plutus-core/plutus-core/src/PlutusCore/DeBruijn/Internal.hs

```diff
@@ -228,19 +238,50 @@ fakeNameDeBruijn (DeBruijn ix) = NamedDeBruijn "i" ix
 nameToDeBruijn
     :: (MonadReader Levels m, AsFreeVariableError e, MonadError e m)
     => Name -> m NamedDeBruijn
-nameToDeBruijn (Name str u) = NamedDeBruijn str <$> getIndex u
+nameToDeBruijn (Name str u) = NamedDeBruijn str <$> getIndex u freeUniqueThrow
 
 tyNameToDeBruijn
     :: (MonadReader Levels m, AsFreeVariableError e, MonadError e m)
     => TyName -> m NamedTyDeBruijn
 tyNameToDeBruijn (TyName n) = NamedTyDeBruijn <$> nameToDeBruijn n
 
 deBruijnToName
-    :: (MonadReader Levels m, AsFreeVariableError e, MonadError e m)
-    => NamedDeBruijn -> m Name
-deBruijnToName (NamedDeBruijn str ix) = Name str <$> getUnique ix
+    :: MonadReader Levels m
+    => (Index -> m Unique)
+    -> NamedDeBruijn
+    -> m Name
+deBruijnToName h (NamedDeBruijn str ix) = Name str <$> getUnique ix h
 
 deBruijnToTyName
-    :: (MonadReader Levels m, AsFreeVariableError e, MonadError e m)
-    => NamedTyDeBruijn -> m TyName
-deBruijnToTyName (NamedTyDeBruijn n) = TyName <$> deBruijnToName n
+    :: MonadReader Levels m
+    => (Index -> m Unique)
+    -> NamedTyDeBruijn
+    -> m TyName
+deBruijnToTyName h (NamedTyDeBruijn n) = TyName <$> deBruijnToName h n
+
+-- | The default handler of throwing an error upon encountering a free name (unique).
+freeUniqueThrow :: (AsFreeVariableError e, MonadError e m) => Unique -> m Index
+freeUniqueThrow =
+    throwing _FreeVariableError . FreeUnique
+
+-- | The default handler of throwing an error upon encountering a free debruijn index.
+freeIndexThrow :: (AsFreeVariableError e, MonadError e m) => Index -> m Unique
+freeIndexThrow =
+    throwing _FreeVariableError . FreeIndex
+
+{-| A different implementation of a handler,  where "free" debruijn indices do not throw an error
+but are instead gracefully converted to fresh uniques.
+These generated uniques remain free; i.e.  if the original term was open, it will remain open after applying this handler.
+These generated free uniques are consistent across the open term (by using a state cache).
+-}
+freeIndexGrace :: (MonadReader Levels m, MonadState (BM.Bimap Unique Level) m, MonadQuote m) => Index -> m Unique
```

Comment:
Naming is hard... "grace" is a bit generic. How about `freeIndexAsConsistentLevel`?

Obviously this would then conflict with your naming of the derived functions... which is partly why I suggested only having the `*With` variants! Not so bad to write `unDeBruijnTermWith freeIndexAsConsistentLevel`